eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala

```diff
@@ -784,54 +825,84 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
+              stay() using d.copy(spliceStatus = SpliceStatus.QuiescenceRequested(cmd))
             }
-          } else {
-            log.warning("cannot initiate splice, channel is not idle or peer doesn't support splices")
-            cmd.replyTo ! RES_FAILURE(cmd, CommandUnavailableInThisState(d.channelId, "splice", NORMAL))
+          case SpliceStatus.NoSplice if !d.commitments.params.useQuiescence =>
+            initiateSplice(cmd, d) match {
+              case Left(f) =>
+                cmd.replyTo ! RES_FAILURE(cmd, f)
+                stay()
+              case Right(spliceInit) =>
+                stay() using d.copy(spliceStatus = SpliceStatus.SpliceRequested(cmd, spliceInit)) sending spliceInit
+            }
+          case _ =>
+            log.warning("cannot initiate splice, another one is already in progress")
+            cmd.replyTo ! RES_FAILURE(cmd, InvalidSpliceAlreadyInProgress(d.channelId))
             stay()
+        }
+      } else {
+        log.warning("cannot initiate splice, peer doesn't support splices")
+        cmd.replyTo ! RES_FAILURE(cmd, CommandUnavailableInThisState(d.channelId, "splice", NORMAL))
+        stay()
+      }
+
+    case Event(msg: Stfu, d: DATA_NORMAL) =>
+      if (d.commitments.params.useQuiescence) {
+        if (d.commitments.remoteIsQuiescent) {
+          d.spliceStatus match {
+            case SpliceStatus.NoSplice =>
+              startSingleTimer(QuiescenceTimeout.toString, QuiescenceTimeout(peer), nodeParams.channelConf.quiescenceTimeout)
+              if (d.commitments.localIsQuiescent) {
+                stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent) sending Stfu(d.channelId, initiator = false)
+              } else {
+                stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
+              }
+            case SpliceStatus.QuiescenceRequested(_) =>
+              stay() using d.copy(spliceStatus = SpliceStatus.ReceivedStfu(msg))
+            case SpliceStatus.InitiatorQuiescent(cmd) =>
+              // if both sides send stfu at the same time, the quiescence initiator is the channel initiator
+              if (!msg.initiator || d.commitments.params.localParams.isInitiator) {
+                initiateSplice(cmd, d) match {
+                  case Left(f) =>
+                    cmd.replyTo ! RES_FAILURE(cmd, f)
+                    context.system.scheduler.scheduleOnce(2 second, peer, Peer.Disconnect(remoteNodeId))
+                    stay() using d.copy(spliceStatus = SpliceStatus.NoSplice) sending Warning(d.channelId, f.getMessage)
+                  case Right(spliceInit) =>
+                    stay() using d.copy(spliceStatus = SpliceStatus.SpliceRequested(cmd, spliceInit)) sending spliceInit
+                }
+              } else {
+                stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent)
```

Comment:
We should also report the failure and log something here, I'm curious to see if this race condition happens frequently in practice:

```suggestion
                log.warning("concurrent stfu received and our peer is the channel initiator, cancelling our splice attempt")
                cmd.replyTo ! RES_FAILURE(cmd, ConcurrentRemoteSplice(d.channelId))
                stay() using d.copy(spliceStatus = SpliceStatus.NonInitiatorQuiescent)
```

Reply:
fixed in [a31a96f](https://github.com/ACINQ/eclair/pull/2680/commits/a31a96f3d756fdc8919fae6c363cf90757047b14).